lib/lurker/templates/lurker/rendering/_submit_form.html.erb

```diff
@@ -6,100 +6,145 @@
-      state.curl = this.renderCurl(method, host, template, values, jQuery.param(JSON.parse('<%= @post_params.to_json.html_safe %>')));
       return state;
     },
-    getPayload: function() {
-      // FIXME
-      return jQuery('#post-payload input').serialize();
-    },
-    getHostname: function(host) {
-      return host;
-    },
-    renderAction: function(host, template, values) {
-      var action = template;
-      var i;
-      for (i in values) {
+    renderAction: function() {
+      var action = this.state.template;
+      var values = this.state.values;
+      for (var i = 0; i < values.length; i++) {
         action = action.replace(new RegExp(':' + values[i].label), values[i].value);
       }
-      return this.getHostname(host) + action;
+      return this.state.host + action;
     },
-    renderCurl: function(method, host, template, values, payload) {
-      var url = '"' + this.renderAction(host, template, values) + '"';
+    renderPayload: function() {
+      var contentType = this.state.requestMediaType;
+      if (contentType === 'application/json' || /\+json$/.test(contentType)) {
+        return JSON.stringify(this.state.payload);
+      } else {
+        // default to 'application/x-www-form-urlencoded'
+        return jQuery.param(this.state.payload);
+      }
+    },
+    renderCurl: function() {
+      var url = this.renderAction();
+      var payload = this.renderPayload();
+
+      var results = [];
+      results.push('-X ' + this.state.method);
       if (payload.length > 0) {
-        url = '-d "' + (payload || this.getPayload()) + '" ' + url;
+        results.push('-H "Content-Type: ' + this.state.requestMediaType + '"');
+        results.push('-d "' + payload.replace(/"/g, '\\"') + '"');
       }
-      return 'curl -X ' + method + " " + url;
+      return 'curl ' + (results.join(' ')) + ' "' + url + '"';
     },
-    // see react-catalyst.js
-    afterSetPartialState: function() {
-      this.setState({
-        action: this.renderAction(this.state.host, this.state.template, this.state.values),
-        curl: this.renderCurl(this.state.method, this.state.host, this.state.template, this.state.values, jQuery('#post-payload input').serialize())
-      });
+    renderHostOptions: function() {
+      var hostOptions = [];
+      for (var name in this.state.hosts) {
+        hostOptions.push(<option key={name} value={this.state.hosts[name]}>{name}</option>);
+      }
+      return hostOptions;
+    },
+    renderRequestMediaTypeOptions: function () {
+      var mediaTypes = this.state.requestMediaTypes;
+      var requestMediaTypeOptions = [];
+      for (var i = 0; i < mediaTypes.length; i++) {
+        requestMediaTypeOptions.push(<option key={i}>{mediaTypes[i]}</option>)
+      }
+      return requestMediaTypeOptions;
     },
     handleSubmit: function() {
-      // FIXME
-      Lurker.onSubmit(jQuery('#payload'));
+      Lurker.disableSubmitButton();
```

Comment:
We need to pass `action`, `payload`, `requestMediaType` and `method`. And it looks a little bit dirty to me. I don't like passing `state` either, since it's an internal implementation detail of the React component. Finally, since request and handler callback are only used from that React component it seemed logical to me to move the code to the component itself.

If you think that it's better to move this code out of the component, I can certainly do that.